File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/AudioFrame.java
package inc.reactor.sdk;

import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;

/**
 * One decoded audio frame: interleaved 16-bit PCM.
 *
 * <p>Same lifetime rule as {@link VideoFrame}: the samples belong to the FFI and are valid only
 * while the handler runs. Reading them afterwards throws {@link IllegalStateException}; {@link
 * #toShortArray()} is the copy.
 */
public final class AudioFrame {

    private final MemorySegment samples;
    private final int sampleCount;
    private final int sampleRate;
    private final int channels;

    AudioFrame(MemorySegment samples, int sampleCount, int sampleRate, int channels) {
        this.samples = samples;
        this.sampleCount = sampleCount;
        this.sampleRate = sampleRate;
        this.channels = channels;
    }

    /**
     * The interleaved samples.
     *
     * <p>Reading this after the handler has returned throws {@link IllegalStateException}.
     *
     * @return a view of the frame's memory
     */
    public MemorySegment samples() {
        return samples;
    }

    /** @return the total number of samples, across every channel */
    public int sampleCount() {
        return sampleCount;
    }

    /** @return samples per second */
    public int sampleRate() {
        return sampleRate;
    }

    /** @return how many channels the samples are interleaved across */
    public int channels() {
        return channels;
    }

    /**
     * Copies the samples onto the Java heap.
     *
     * @return a copy that outlives the handler
     */
    public short[] toShortArray() {
        return samples.toArray(ValueLayout.JAVA_SHORT);
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/AudioFrame.kt
package inc.reactor.sdk.android

import java.nio.ShortBuffer

/**
 * One decoded audio frame: interleaved signed 16-bit PCM.
 *
 * **[samples] is borrowed**, exactly as [VideoFrame.pixels] is — valid only until the handler
 * returns. Roughly 10 ms of audio arrives per frame.
 */
public class AudioFrame internal constructor(
    public val trackName: String,
    /** Borrowed interleaved int16 PCM. Valid only for this call. */
    public val samples: ShortBuffer,
    /** Total samples across all channels — not frames-per-channel. */
    public val sampleCount: Int,
    public val sampleRate: Int,
    public val channels: Int,
) {
    /** A copy of [samples] that outlives the callback. */
    public fun copySamples(): ShortArray {
        val duplicate = samples.duplicate()
        return ShortArray(duplicate.remaining()).also { duplicate.get(it) }
    }

    override fun toString(): String = "AudioFrame($trackName, $sampleCount samples, ${sampleRate}Hz, ${channels}ch)"
}
